src/wrapper/class_loader.rs

```diff
@@ -0,0 +1,59 @@
+use std::sync::Mutex;
+
+use errors::*;
+
+use JNIEnv;
+
+use objects::{GlobalRef, JObject, JClass, JValue};
+
+use strings::JNIString;
+
+use lazy_static::lazy_static;
+
+/// The `loadClass` function name.
+const LOAD_CLASS: &str = "loadClass";
+/// The `loadClass` signature.
+const LOAD_CLASS_SIG: &str = "(Ljava/lang/String;)Ljava/lang/Class;";
+
+lazy_static! {
+    /// The global class loader instance.
+    static ref CLASS_LOADER: Mutex<Option<GlobalRef>> = Mutex::default();
+}
+
+/// Register the given object as the global class loader instance.
+pub fn register_class_loader<'a>(env: &JNIEnv<'a>, class_loader: JObject<'a>) -> Result<()> {
+    // Check that the `loadClass` function is present.
+    env.get_method_id(class_loader, LOAD_CLASS, LOAD_CLASS_SIG)?;
+
+    *CLASS_LOADER.lock().unwrap() = Some(env.new_global_ref(class_loader)?);
+
+    Ok(())
+}
+
+/// Unregister the global class loader instance.
+pub fn unregister_class_loader() {
+    *CLASS_LOADER.lock().unwrap() = None;
+}
+
+/// Look up a class by name.
+///
+/// Either it uses the registered `CLASS_LOADER` or it falls back to use the
+/// JNI env function `FindClass`.
+pub(crate) fn load_class<'a>(env: &JNIEnv<'a>, name: JNIString) -> Result<JClass<'a>> {
+    match *CLASS_LOADER.lock().unwrap() {
```

Comment:
This code serializes all class lookups. I think we must ensure that normal `find_class` clients do not pay any price for this feature and are not serialized.

Possibly, one way to achieve that is to acquire the lock only to read the optional because there is no requirement of a mutual exclusion between the class lookups and `register_class_loader`/`unregister_class_loader` (i.e., it is totally OK to let "the last" lookup to proceed). GlobalRefs can be cloned and safely accessed from multiple threads — the last one dropping it will free it.

Reply:
I'm wouldn't it be more optimal to use `RwLock` instead of `Mutex` here? 🤔

Reply:
To read a couple of words, [unlikely](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/locks/ReadWriteLock.html) — but the only way to definitely tell is to measure. The problem is coming with a _realistic_ scenario — the worst-case one we can easily put together.